### PR TITLE
Chores Post Release

### LINKED ISSUE #150: Chores post-release 10.2.0.0
#### Description
Quick updates and fixes now that the release v10.2.0.0 is out.

#### Questions/Ideas
- [x] Update dev docker image to khiops 10.2.0
- ~~Make the API docs show the Khiops version in the top left side. The `khiops-python` version must be shown somewhere else in the index page.~~ We'll just keep the KP version.
- [x] Update CI
  - [x] Update checkout and artifacts actions to v4
    - The artifacts actions changed API: https://github.com/actions/upload-artifact/blob/main/docs/MIGRATION.md
  - [x] Simplify workflow names so they are better seen in PRs
     - 1 or 2 words workflow names
     - jobs can have longer names
  - Simplify CI:
     - [x] ~~Join conda and pip packaging into a packaging workflow~~ on second thought it is better to have them separated.
  - [x] Set `main` as default value for `KHIOPS_REVISION` in the conda workflow.
  - [x] Replace the `push: tags: [v*]` to `push: tags:`
  - [x] Make the pages deploy only manual.
- Warning and messages:
  - [x] Make the "old" API version appear once at load time instead of at each API call
  - [x] Make the failure messages of the runner be more user friendly

### PR DESCRIPTION
**Pull request by folmos-at-orange**
Closes #150

### CODE REVIEW

File: .github/workflows/api-docs.yml [hunk 1]
```diff
@@ -24,21 +29,20 @@ concurrency:
   group: pages
   cancel-in-progress: false
 jobs:
-  create-docs:
-    name: Create Sphinx docs
-    runs-on: ubuntu-22.04
+  build:
+    runs-on: ubuntu-latest
```
Comment (popescu-v) [inline .github/workflows/api-docs.yml:33]: Why `ubuntu-latest` here, while we have `ubuntu-22.04` elsewhere? Is there a reason the same version specification cannot be used consistently throughout the workflows?

File: .github/workflows/dev-docker.yml [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+          tags: ghcr.io/khiopsml/khiops-python/khiopspydev-${{ matrix.khiopsdev-os }}:latest
+          # Push only on manual request
+          push: ${{ inputs.push || false }}
+      - name: Display the image digest
+        run: echo ${{ steps.docker-build.outputs.digest }}
+  test:
+    # Test only when manually pushed an requested
+    if: ${{ inputs.push || false }}
+    needs: build-push
+    runs-on: ubuntu-22.04
```
Comment (popescu-v) [inline .github/workflows/dev-docker.yml:72]: Version specifier change (see above points in this respect)?

File: .github/workflows/pip.yml [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        description: khiops-samples repo revision
+  pull_request:
+  push:
+    tags: ['*']
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
+  cancel-in-progress: true
+jobs:
+  build:
+    runs-on: ubuntu-22.04
```
Comment (popescu-v) [inline .github/workflows/pip.yml:19]: Version specifier seems OK here (as the code is run directly, not through a container).

File: .github/workflows/pip.yml [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+          # Build the package
+          python3 setup.py sdist
+      - name: Upload package as artifact
+        uses: actions/upload-artifact@v4
+        with:
+          name: pip-package
+          path: ./dist/khiops*.tar.gz
+  test:
+    needs: build
+    runs-on: ubuntu-22.04
```
Comment (popescu-v) [inline .github/workflows/pip.yml:46]: Why this version specifier, as the code is run in a container anyway? Why is `-latest` not enough? Or is it?

File: .github/workflows/unit-tests.yml [hunk 5]
```diff
@@ -18,8 +18,8 @@ concurrency:
   group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
   cancel-in-progress: true
 jobs:
-  run-unit-tests:
-    runs-on: ubuntu-latest
+  run:
+    runs-on: ubuntu-22.04
```
Comment (popescu-v) [inline .github/workflows/unit-tests.yml:22]: Version specifier OK here, as  the code is run directly (not in a container), right?

File: khiops/core/internals/runner.py [hunk 6]
```diff
@@ -1251,6 +1128,56 @@ def _initialize_mpi_command_args(self):
...
+                "-n",
+                str(self.max_cores + 1),
+            ]
+        elif platform.system() == "Darwin":
+            # Note: The '-host localhost' arguments for arm64
+            #       may be removed when mpich > 4.1.2 is released
+            if platform.processor() == "arm":
+                self.mpi_command_args += [
+                    "-host",
+                    "localhost",
```
Comment (popescu-v) [inline khiops/core/internals/runner.py:1159]: Do we still need these, or can we update the `mpich` version used?

File: khiops/core/internals/runner.py [hunk 7]
```diff
@@ -1251,6 +1128,56 @@ def _initialize_mpi_command_args(self):
...
+            self.mpi_command_args += [
+                "-al",
+                "spr:P",
+                "-n",
+                str(self.max_cores + 1),
+                "/priority",
+                "1",
+            ]
+        else:
+            raise KhiopsRuntimeError(f"System '{platform.system()}' not supported.")
```
Comment (popescu-v) [inline khiops/core/internals/runner.py:1179]: Add mention of the khiops.org web site, to check the supported OSes?

File: khiops/core/internals/runner.py [hunk 8]
```diff
@@ -1314,31 +1292,51 @@ def _initialize_default_system_wide_khiops_bin_dir(self):
...
+        self._khiops_version = KhiopsVersion(khiops_version_str)
+
+        # Warn if the khiops version is too far from the khiops-python version
+        compatible_khiops_version = khiops.get_compatible_khiops_version()
+        if self._khiops_version.major > compatible_khiops_version.major:
             warnings.warn(
-                "Sample datasets location is not a directory "
-                f"({samples_dir_to_check}). {download_msg}",
+                f"Khiops version '{self._khiops_version}' is ahead "
+                f"from the khiops-python version '{khiops.__version__}'. "
```
Comment (popescu-v) [inline khiops/core/internals/runner.py:1336]: s/"ahead from"/"ahead of"/

File: packaging/docker/khiopspydev/Dockerfile.rocky [hunk 9]
```diff
@@ -11,29 +11,31 @@ LABEL description="Container for the development of khiops-python"
...
   && source /etc/profile.d/modules.sh \
   && module unload mpi \
   # Hard-code MPICH module name \
   && module load mpi/mpich-x86_64 \
-  && cmake --preset linux-gcc-release -DTESTING=OFF -DBUILD_JARS=OFF -DCMAKE_INSTALL_PREFIX= \
+  && cmake --preset linux-gcc-release -DTESTING=OFF -DBUILD_JARS=ON -DCMAKE_INSTALL_PREFIX= \
   && cmake --build --preset linux-gcc-release --parallel \
-       --target MODL${MPI_SUFFIX} MODL_Coclustering${MPI_SUFFIX} KhiopsNativeInterface \
+       --target MODL${MPI_SUFFIX} MODL_Coclustering${MPI_SUFFIX} \
+                KhiopsNativeInterface norm_jar khiops_jar \
```
Comment (popescu-v) [inline packaging/docker/khiopspydev/Dockerfile.rocky:33]: Why do we need the JARs here (khiops-python dev container)?

File: packaging/docker/khiopspydev/Dockerfile.rocky [hunk 10]
```diff
@@ -11,29 +11,31 @@ LABEL description="Container for the development of khiops-python"
...
   && cmake --install ./build/linux-gcc-release \
   && cd .. \
   # Set python to python3 \
   && alternatives --install /usr/bin/python python /usr/bin/python3 1 \
   # Clean build files \
-  && yum clean all \
```
Comment (popescu-v) [inline packaging/docker/khiopspydev/Dockerfile.rocky:35]: Add ```sh
&& dnf clean all
```
to keep the image small(er)?

File: packaging/docker/khiopspydev/Dockerfile.ubuntu [hunk 11]
```diff
@@ -14,12 +14,12 @@ RUN true \
   && apt-get -y update \
   && apt-get -y --no-install-recommends install git python3-pip zip pandoc \
   # Obtain, build and install Khiops \
-  && git clone --depth 1 https://github.com/khiopsml/khiops.git \
+  && git clone https://github.com/khiopsml/khiops.git \
   && cd khiops \
   && git checkout ${KHIOPS_REVISION} \
-  && cmake --preset linux-gcc-release -DTESTING=OFF -DBUILD_JARS=OFF -DCMAKE_INSTALL_PREFIX= \
+  && cmake --preset linux-gcc-release -DTESTING=OFF -DBUILD_JARS=ON -DCMAKE_INSTALL_PREFIX= \
```
Comment (popescu-v) [inline packaging/docker/khiopspydev/Dockerfile.ubuntu:20]: Why do we need the JARs on these khiops-python dev containers?

File: packaging/docker/khiopspydev/Dockerfile.ubuntu [hunk 12]
```diff
@@ -14,12 +14,12 @@ RUN true \
...
   && cmake --build --preset linux-gcc-release --parallel \
-       --target MODL MODL_Coclustering KhiopsNativeInterface \
+       --target MODL MODL_Coclustering KhiopsNativeInterface norm_jar khiops_jar \
```
Comment (popescu-v) [inline packaging/docker/khiopspydev/Dockerfile.ubuntu:22]: Why do we need the JARs on these khiops-python dev containers?

File: .github/workflows/api-docs.yml (hunk 1 above)
Comment (folmos-at-orange) [reply .github/workflows/api-docs.yml:33]: I will change to `ubuntu-22.04`

File: .github/workflows/dev-docker.yml (hunk 2 above)
Comment (folmos-at-orange) [reply .github/workflows/dev-docker.yml:72]: See above

File: .github/workflows/pip.yml (hunk 3 above)
Comment (folmos-at-orange) [reply .github/workflows/pip.yml:19]: See above

File: .github/workflows/pip.yml (hunk 4 above)
Comment (folmos-at-orange) [reply .github/workflows/pip.yml:46]: See above. 

And I `latest` will change suddenly this year. To not have suprises, I prefer migrate it explicitly when the  `ubuntu-24.04` will be available.

File: .github/workflows/unit-tests.yml (hunk 5 above)
Comment (folmos-at-orange) [reply .github/workflows/unit-tests.yml:22]: See above.

File: khiops/core/internals/runner.py (hunk 6 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:1159]: Maybe later, when we bump the MPI version.

File: khiops/core/internals/runner.py (hunk 8 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:1336]: Ok

File: packaging/docker/khiopspydev/Dockerfile.rocky [hunk 13]
```diff
@@ -11,29 +11,31 @@ LABEL description="Container for the development of khiops-python"
...
+  && git clone https://github.com/khiopsml/khiops.git \
   && cd khiops \
   && git checkout ${KHIOPS_REVISION} \
   # Make sure that MPI is mpich \
   && source /etc/profile.d/modules.sh \
   && module unload mpi \
   # Hard-code MPICH module name \
   && module load mpi/mpich-x86_64 \
-  && cmake --preset linux-gcc-release -DTESTING=OFF -DBUILD_JARS=OFF -DCMAKE_INSTALL_PREFIX= \
+  && cmake --preset linux-gcc-release -DTESTING=OFF -DBUILD_JARS=ON -DCMAKE_INSTALL_PREFIX= \
```
[the remark this answers is not in the document]
Comment (folmos-at-orange) [reply packaging/docker/khiopspydev/Dockerfile.rocky:30]: This setting was `ON` by default before. Without the JARs `cmake` install fails.

File: packaging/docker/khiopspydev/Dockerfile.rocky (hunk 9 above)
Comment (folmos-at-orange) [reply packaging/docker/khiopspydev/Dockerfile.rocky:33]: See above

File: .github/workflows/api-docs.yml (hunk 1 above)
Comment (folmos-at-orange) [reply .github/workflows/api-docs.yml:33]: Done

File: packaging/docker/khiopspydev/Dockerfile.rocky (hunk 13 above)
[the remark this answers is not in the document]
Comment (popescu-v) [reply packaging/docker/khiopspydev/Dockerfile.rocky:30]: So this is is to account for a change in the Khiops `cmake` scripts, right? As this used to work before and was more intuitive IMHO, and adequate to the narrower scope that we have here (a dev container for just the binaries). If this is the case, then a comment would be welcome IMHO.

File: packaging/docker/khiopspydev/Dockerfile.rocky (hunk 9 above)
Comment (popescu-v) [reply packaging/docker/khiopspydev/Dockerfile.rocky:33]: OK, pending relevant comment documenting the upstream change which requires this modification here.

File: packaging/docker/khiopspydev/Dockerfile.rocky (hunk 13 above)
[the remark this answers is not in the document]
Comment (folmos-at-orange) [reply packaging/docker/khiopspydev/Dockerfile.rocky:30]: Ok

File: packaging/docker/khiopspydev/Dockerfile.rocky (hunk 10 above)
Comment (folmos-at-orange) [reply packaging/docker/khiopspydev/Dockerfile.rocky:35]: Ok

File: packaging/docker/khiopspydev/Dockerfile.ubuntu (hunk 11 above)
Comment (folmos-at-orange) [reply packaging/docker/khiopspydev/Dockerfile.ubuntu:20]: See above

File: packaging/docker/khiopspydev/Dockerfile.ubuntu (hunk 12 above)
Comment (folmos-at-orange) [reply packaging/docker/khiopspydev/Dockerfile.ubuntu:22]: See above.

File: khiops/core/internals/runner.py (hunk 8 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:1336]: Done

File: khiops/core/internals/runner.py (hunk 7 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:1179]: Done

File: packaging/docker/khiopspydev/Dockerfile.rocky (hunk 13 above)
[the remark this answers is not in the document]
Comment (folmos-at-orange) [reply packaging/docker/khiopspydev/Dockerfile.rocky:30]: Done

File: packaging/docker/khiopspydev/Dockerfile.rocky (hunk 9 above)
Comment (folmos-at-orange) [reply packaging/docker/khiopspydev/Dockerfile.rocky:33]: Done

File: packaging/docker/khiopspydev/Dockerfile.rocky (hunk 10 above)
Comment (folmos-at-orange) [reply packaging/docker/khiopspydev/Dockerfile.rocky:35]: Done

File: packaging/docker/khiopspydev/Dockerfile.ubuntu (hunk 11 above)
Comment (folmos-at-orange) [reply packaging/docker/khiopspydev/Dockerfile.ubuntu:20]: Ok with comment and  done.

File: khiops/core/internals/runner.py [hunk 14]
```diff
@@ -1251,6 +1134,59 @@ def _initialize_mpi_command_args(self):
...
+                "spr:P",
+                "-n",
+                str(self.max_cores + 1),
+                "/priority",
+                "1",
+            ]
+        else:
+            raise KhiopsEnvironmentError(
+                f"Unsupported OS {platform.system()}. "
+                "Check the supported OS's at https://khiops.org."
```
Comment (popescu-v) [inline khiops/core/internals/runner.py:1187]: s/OS's/OSes/

File: khiops/core/internals/runner.py [hunk 15]
```diff
@@ -114,7 +114,10 @@ def _get_system_cpu_cores():
     elif platform.system() == "Darwin":
         cpu_system_info_args = ["sysctl", "-n", "hw.physicalcpu"]
     else:
-        raise KhiopsRuntimeError(f"The '{platform.system()}' is not supported.")
+        raise KhiopsEnvironmentError(
+            f"Unsupported OS {platform.system()}. "
+            "Check the supported OS's at https://khiops.org."
```
Comment (popescu-v) [inline khiops/core/internals/runner.py:119]: s/OS'S/OSes

File: khiops/core/internals/runner.py [hunk 16]
```diff
@@ -144,7 +147,10 @@ def _get_system_cpu_cores():
     elif platform.system() == "Darwin":
         cpu_core_count = int(cpu_system_info_output.strip())
     else:
-        raise KhiopsRuntimeError(f"The '{platform.system()}' is not supported.")
+        raise KhiopsEnvironmentError(
+            f"Unsupported OS {platform.system()}. "
+            "Check the supported OS's at https://khiops.org."
```
Comment (popescu-v) [inline khiops/core/internals/runner.py:152]: s/OS's/OSes

File: khiops/core/internals/runner.py [hunk 17]
```diff
@@ -1314,31 +1301,51 @@ def _initialize_default_system_wide_khiops_bin_dir(self):
             self._khiops_bin_dir = os.path.join(os.path.sep, "usr", "bin")
         # Raise an error for unknown platforms
         else:
-            raise KhiopsEnvironmentError(f"Unsupported platform {platform.system()}")
+            raise KhiopsEnvironmentError(
+                f"Unsupported OS {platform.system()}. "
+                "Check the supported OS's at https://khiops.org."
```
Comment (popescu-v) [inline khiops/core/internals/runner.py:1306]: s/OS's/OSes/ (everywhere applicable)

**Review by popescu-v — APPROVED** [review]
LGTM (modulo one really minor spelling comment).

File: khiops/core/internals/runner.py (hunk 14 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:1187]: Ok

File: khiops/core/internals/runner.py (hunk 15 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:119]: Ok

File: khiops/core/internals/runner.py (hunk 16 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py:152]: Ok

**Pull request merged**